File: cpp/problem558.cpp
#include <mystd.h>
using namespace std;
using namespace MYSTD;


/**
 * definition for a QuadTree node.
 */
class Node {
public:
    bool val;
    bool isLeaf;
    Node* topLeft;
    Node* topRight;
    Node* bottomLeft;
    Node* bottomRight;

    Node() {}
    Node(bool _val, bool _isLeaf, Node* _topLeft, Node* _topRight, Node* _bottomLeft, Node* _bottomRight) {
        val = _val;
        isLeaf = _isLeaf;
        topLeft = _topLeft;
        topRight = _topRight;
        bottomLeft = _bottomLeft;
        bottomRight = _bottomRight;
    }
};


/**
 * Quad Tree - full implementation of intersect(), with some
 * helping ultilities for construting, destroying and debugging.
 */
class QuadTree{
public:
    /**
     * inializing of Quad Tree
     * If you use a value 'val' to initialize it, this generate a new tree.
     * If use a root node 'tr', the class takes over the tree, and is responsible
     * for the memory management.
     */
    QuadTree( bool val ):_root(copyLeaf(val)){ }
    QuadTree( Node *tr ):_root(tr){}
    ~QuadTree(){ delSubTree(_root); }
    /**
     * solution to the question intersect()
     */
    static Node * intersect(Node* tr1, Node* tr2) {
        if( tr1->isLeaf ){
            if( tr2->isLeaf ) return intersectLeafLeaf( tr1, tr2 );
            else return intersectBranchLeaf( tr2, tr1 );
        }else{
            if( tr2->isLeaf ) return intersectBranchLeaf( tr1, tr2 );
            else return intersectBranchBranch( tr1, tr2 );
        }
    }
    /**
     * print the tree nodes
     */
    ostream & print( ostream &os = cout ){
        return print( _root, os, 0 );
    }
    /**
     * return root node of the tree
     */
    Node * root(){
        return _root;
    }
    /**
     * extend a leaf node to a subtree
     */
    static void extend(Node *leaf, int val1, int val2, int val3, int val4){
        leaf->isLeaf = false;
        leaf->topLeft = copyLeaf( val1 );
        leaf->topRight = copyLeaf( val2 );
        leaf->bottomLeft = copyLeaf( val3 );
        leaf->bottomRight = copyLeaf( val4 );
    }
protected:
    Node *_root;

    static Node * copySubTree(Node *tr) {
        if( !tr ) return nullptr;
        return new Node( tr->val, tr->isLeaf, 
            copySubTree( tr->topLeft ), copySubTree( tr->topRight ), 
            copySubTree( tr->bottomLeft ), copySubTree( tr->bottomRight ) );
    }
    static Node * copyLeaf( bool val ){
        return new Node( val, true, nullptr, nullptr, nullptr, nullptr );
    }
    static void delSubTree(Node *&tr) {
        if( !tr ) return;
        delSubTree(tr->topLeft); delSubTree(tr->topRight);
        delSubTree(tr->bottomLeft); delSubTree(tr->bottomRight);
        delete tr;
        tr = nullptr;
    }
    static Node * intersectLeafLeaf(Node *tr1, Node *tr2){
        return copyLeaf( tr1->val || tr2->val );
    }
    static Node * intersectBranchLeaf(Node *tr1, Node *tr2){
        return tr2->val ? copyLeaf(true) : copySubTree(tr1);
    }
    static Node * intersectBranchBranch(Node *tr1, Node *tr2){
        auto *tr = new Node(false, false, 
            intersect( tr1->topLeft, tr2->topLeft ),
            intersect( tr1->topRight, tr2->topRight ),
            intersect( tr1->bottomLeft, tr2->bottomLeft ),
            intersect( tr1->bottomRight, tr2->bottomRight ) );
        bool val = tr->topLeft->val;
        if( tr->topLeft->isLeaf && tr->topRight->isLeaf && 
            tr->bottomLeft->isLeaf && tr->bottomRight->isLeaf && 
            tr->topRight->val == val && tr->bottomLeft->val == val &&
            tr->bottomRight->val == val ){
            delSubTree( tr->topLeft);
            delSubTree( tr->topRight);
            delSubTree( tr->bottomLeft);
            delSubTree( tr->bottomRight );
            tr->val = val;
            tr->isLeaf = true;
        }
        return tr;
    }
    static ostream & print( Node *tr, ostream &os = cout, int indent = 0 ){
        if( !tr ) return os;
        if( tr->isLeaf ){
            os << tr->val << endl;
        }else{
            os << endl;
            os << string(indent, ' ') << "topLeft: "; print( tr->topLeft, os, indent+4 );
            os << string(indent, ' ') << "topRight: "; print( tr->topRight, os, indent+4 );
            os << string(indent, ' ') << "bottomLeft: "; print( tr->bottomLeft, os, indent+4 );
            os << string(indent, ' ') << "bottomRight: "; print( tr->bottomRight, os, indent+4 );
        }
        return os;
    }
};



/**
 * solution with recursive merging, copy, and delete
 * Time cost O(n), Space cost O(n), where is number of nodes.
 *
 * Note that some C++ implementations in Leetcode discussion and solution
 * has severe logic problem (they mix the deep copy and shallow copy without
 * safe memory release).
 *
 * Author: Yangyao, 2019/01/16
 */
class Solution {
public:
    Node * intersect(Node* quadTree1, Node* quadTree2) {
        return QuadTree::intersect(quadTree1, quadTree2);
    }
};

int main(int argc, char const *argv[]){
    QuadTree tr1(true);
    auto *p1 = tr1.root();
    tr1.extend(p1, 1, 1, 0, 0);
    
    QuadTree tr2(true);
    auto *p2 = tr2.root();
    tr2.extend(p2, 1, 1, 1, 0);
    tr2.extend(p2->topRight, 0, 0, 1, 1);

    tr1.print();
    tr2.print();

    Solution sol;
    QuadTree tr12( sol.intersect( p1, p2 ) );
    tr12.print();

    return 0;
}

File: java/problem558.java
/*
update (using methond combineSameNode())
56 / 58 test cases passed. 
*/

/*
!!!!!!!!!bug still exist!!!!!!!!!!!!!!!! 
leetcode test results: only 27 / 58 test cases passed
*/

import java.util.Stack;
// Definition for a QuadTree node.
class Node {
    public boolean val;
    public boolean isLeaf;
    public Node topLeft;
    public Node topRight;
    public Node bottomLeft;
    public Node bottomRight;

    public Node() {}

    public Node(boolean _val,boolean _isLeaf,Node _topLeft,Node _topRight,Node _bottomLeft,Node _bottomRight) {
        val = _val;
        isLeaf = _isLeaf;
        topLeft = _topLeft;
        topRight = _topRight;
        bottomLeft = _bottomLeft;
        bottomRight = _bottomRight;
    }
};
class Solution {
    public Node intersect(Node quadTree1, Node quadTree2) {
        Stack<Node> newFather=new Stack<>();
        Node root=new Node();
        newFather.push(root);

        Stack<Node> st1=new Stack<Node>();
        Stack<Node> st2=new Stack<Node>();
        st1.push(quadTree1);st2.push(quadTree2);
        while(!st1.empty()){
            Node temp1=st1.peek();
            Node temp2=st2.peek();
            Node fr=newFather.peek();

            if(!temp1.isLeaf&&!temp2.isLeaf){
                newFather.pop();
                for(int i=0;i<4;i++){
                    Node temp=new Node();
                    switch (i) {
                        case 3:
                            fr.topLeft=temp;
                            break;
                        case 2:
                            fr.topRight=temp;
                            break;
                        case 1:
                            fr.bottomLeft=temp;
                            break;
                        case 0:
                            fr.bottomRight=temp;
                            break;
                        default:
                            break;
                    }
                    newFather.push(temp);
                }
                st1.pop();
                st1.push(temp1.bottomRight);
                st1.push(temp1.bottomLeft);
                st1.push(temp1.topRight);
                st1.push(temp1.topLeft);

                st2.pop();
                st2.push(temp2.bottomRight);
                st2.push(temp2.bottomLeft);
                st2.push(temp2.topRight);
                st2.push(temp2.topLeft);      
            }else if(temp1.isLeaf&&!temp2.isLeaf){
                if(temp1.val){
                    fr.isLeaf=true;fr.val=true;
                    st1.pop();st2.pop();newFather.pop();
                }else{
                    newFather.pop();
                    for(int i=0;i<4;i++){
                        Node temp=new Node();
                        switch (i) {
                            case 3:
                                fr.topLeft=temp;
                                break;
                            case 2:
                                fr.topRight=temp;
                                break;
                            case 1:
                                fr.bottomLeft=temp;
                                break;
                            case 0:
                                fr.bottomRight=temp;
                                break;
                            default:
                                break;
                        }
                        newFather.push(temp);
                    }
                    st2.pop();
                    st2.push(temp2.bottomRight);
                    st2.push(temp2.bottomLeft);
                    st2.push(temp2.topRight);
                    st2.push(temp2.topLeft); 
                    
                    st1.pop();
                    st1.push(temp1);
                    st1.push(temp1);
                    st1.push(temp1);
                    st1.push(temp1); 
                }

            }else if(!temp1.isLeaf&&temp2.isLeaf){
                if(temp2.val){
                    fr.isLeaf=true;fr.val=true;
                    st1.pop();st2.pop();newFather.pop();
                }else{
                    newFather.pop();
                    for(int i=0;i<4;i++){
                        Node temp=new Node();
                        switch (i) {
                            case 3:
                                fr.topLeft=temp;
                                break;
                            case 2:
                                fr.topRight=temp;
                                break;
                            case 1:
                                fr.bottomLeft=temp;
                                break;
                            case 0:
                                fr.bottomRight=temp;
                                break;
                            default:
                                break;
                        }
                        newFather.push(temp);
                    }
                    st1.pop();
                    st1.push(temp1.bottomRight);
                    st1.push(temp1.bottomLeft);
                    st1.push(temp1.topRight);
                    st1.push(temp1.topLeft);

                    st2.pop();
                    st2.push(temp2);
                    st2.push(temp2);
                    st2.push(temp2);
                    st2.push(temp2);     
                }

            }else{
                fr.isLeaf=true;fr.val=temp1.val||temp2.val;
                st1.pop();st2.pop();newFather.pop();
            }

        }
        st1.push(root);
        boolean tag=false;;
        while(true){
            if(st1.isEmpty()){
                if(!tag){
                    break;
                }else{
                    st1.push(root);
                }
            }
            tag=false;
            Node temp=st1.peek();
            if(temp!=null){
                tag=combineSameNode(temp);
                st1.pop();
                st1.push(temp.bottomRight);
                st1.push(temp.bottomLeft);
                st1.push(temp.topRight);
                st1.push(temp.topLeft);
            }else{
                st1.pop();
            }
        }
        /*del node with same child*/
        /*
            st1.push(root);
            while(!st1.isEmpty()){
                Node temp=st1.peek();
                if(temp==null){
                    break;
                }else if(!temp.isLeaf
                &&valOfNode(temp.topLeft)==valOfNode(temp.topRight)
                &&valOfNode(temp.topLeft)==valOfNode(temp.bottomLeft)
                &&valOfNode(temp.topLeft)==valOfNode(temp.bottomRight)
                &&valOfNode(temp.topLeft)!=-1
                ){
                    temp.isLeaf=true;
                    if(valOfNode(temp.topLeft)==1){
                        temp.val=true;
                    }else{
                        temp.val=false;
                    }
                    temp.topLeft=null;temp.topRight=null;
                    temp.bottomLeft=null;temp.bottomRight=null;
                    st1.pop();
                }else{
                    st1.pop();
                    st1.push(temp.bottomRight);
                    st1.push(temp.bottomLeft);
                    st1.push(temp.topRight);
                    st1.push(temp.topLeft);
                }
            }
        */

        return root;
    }
    public static boolean combineSameNode(Node node){
        if(node.isLeaf){
            return false;
        }else if(node.bottomLeft.isLeaf
        &&node.bottomRight.isLeaf
        &&node.topLeft.isLeaf
        &&node.topRight.isLeaf){
            if(node.bottomLeft.val==node.bottomRight.val
            &&node.bottomLeft.val==node.topLeft.val
            &&node.bottomLeft.val==node.topRight.val){
                node.isLeaf=true;node.val=node.bottomLeft.val;
                node.topLeft=null;node.topRight=null;
                node.bottomLeft=null;node.bottomRight=null;
                return true;
            }
        }
        return false;
    }
    public static byte valOfNode(Node node){
        if(node.isLeaf){
            if(node.val){
                return 1;
            }else{
                return 0;
            }
        }else{
            if(valOfNode(node.topLeft)==valOfNode(node.topRight)&&valOfNode(node.topLeft)
            ==valOfNode(node.bottomLeft)&&valOfNode(node.topLeft)==valOfNode(node.bottomRight)){
              return valOfNode(node.topLeft); 
            }else{
                return -1;
            }
        }

    }

}

